### PR TITLE
Allow identity transitions

### PR DESCRIPTION
**Pull request by marcosh**
Practically, this is needed to implement the `Choice` typeclass for a state machine.

Since in choice the state machine evolves depending on the input, it could happen that it stays where it is. Therefore we always allow the identity transition, so that such case is always allowed

### CODE REVIEW

File: decision-log/2023-01-13-always-allow-identity-transitions.yaml [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+decision: >
+  We decide to consider everytime not changing state as a valid transition.
+consequences: >
+  We create a new instance of our `AllowedTransition` typeclass which searches
+  for a proof term indicating that a transition is allowed. Such an instance
+  says that going from a vertex to itself in the space of the states of the
+  machine is always allowed.
+
+  This implies that it is not necessary to specify in a topology that every
+  trivial transition is allowed.
```
[the remark this answers is not in the document]
Comment (marcosh) [reply decision-log/2023-01-13-always-allow-identity-transitions.yaml:18]: Definitely makes sense to mention it in the consequences.

Probably there are cases where you don't want the trivial transitions, for for the moment I think the pros of this decision outweigh the cons. If we gather evidence that this decision creates more issues than benefits, we will review it

File: src/CRM/Topology.hs [hunk 2]
```diff
@@ -23,19 +23,39 @@ $( singletons
...
+
+instance {-# INCOHERENT #-} AllowedTransition ('Topology ('(a, l1) : l2)) a b => AllowedTransition ('Topology ('(a, x : l1) : l2)) a b where
+  allowsTransition =
+    AllowAddingEdge (allowsTransition :: AllowTransition ('Topology ('(a, l1) : l2)) a b)
+
+instance {-# INCOHERENT #-} AllowedTransition ('Topology map) a b => AllowedTransition ('Topology (x : map)) a b where
+  allowsTransition =
+    AllowAddingVertex (allowsTransition :: AllowTransition ('Topology map) a b)
+
+instance {-# INCOHERENT #-} AllowedTransition topology a a where
```
[the remark this answers is not in the document]
Comment (marcosh) [reply src/CRM/Topology.hs:60]: I guess you're asking about something like [this](https://github.com/tweag/crem/blob/8d3bf47bc8cfe8a90afce3f69d48cb5c50878dcf/src/CRM/Topology.hs#L65), which will arrive in a future PR.

I wouldn't move it to this branch, if you don't mind

[the remark this answers is not in the document]
Comment (AlexeiDrake) [reply src/CRM/Topology.hs:60]: I agree we shouldn't mix up the order of the PRs.

That `TrivialTopology` is not quite what I had in mind; that one describes no allowed state transitions other than trivial ones.

I'm thinking about having multiple types of topologies:
- one in which only the given transitions are allowed
- one in which the given ∪ identity transitions are allowed
- one in which every possible transition is allowed, in a way that does not store every allowed transition to allow for large sets, e.g. `Int`

To achieve this, we could change `Topology` into a typeclass or a GADT.

File: decision-log/2023-01-13-always-allow-identity-transitions.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (AlexeiDrake) [reply decision-log/2023-01-13-always-allow-identity-transitions.yaml:18]: Another consideration is that the current `Topology` model does not scale well to really large fully-connected graphs, as described below and in #26.

File: src/CRM/Topology.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (AlexeiDrake) [reply src/CRM/Topology.hs:60]: I'll open an issue about this, and unblock these two PRs.

**Review by AlexeiDrake — APPROVED** [review]

**Pull request merged**